### PR TITLE
Concatenate to support Ellipsis argument

### PR DESCRIPTION
**Pull request by Daraan**
Fixes https://github.com/python/typing_extensions/issues/110.

This PR adds support for Ellipsis in Concatenate for the Python 3.8-3.10 backports.

Does not fix:
   - https://github.com/python/typing_extensions/issues/48 : Using Ellipsis as argument for Callable with a generic..
      -  <= 3.9: because of typing.Callable limitations
      -  For 3.10 needs modification of `_ConcatenateGenericAlias`

Changes:
    - Changed _concatenate_getitem to reflect the Python3.11 typing module that allows the Ellipsis parameter.
    - For Python 3.10 uses typing._ConcatenateGenericAlias and _concatenate_getitem
    - For <=Python 3.9 the typing_extensions._ConcatenateGenericAlias and _concatenate_getitem like before
    - For <3.9.2 uses a Dummy class that is temporarily swapped with an Ellipsis input
    - added tests and updated tests

### CODE REVIEW

**Comment by Daraan** [comment]
I am sorry I had to recreate #442.

Latest change since then:

```diff
-        if not TYPING_3_11_0:
-            with self.assertRaisesRegex(
-                TypeError,
-                'each arg must be a type',
-           ):
-              Concatenate[1, P]
# different location
+    def test_invalid_use(self):
+       # Assure that `_type_check` is called.
+        P = ParamSpec('P')
+        with self.assertRaisesRegex(
+            TypeError,
+            "each arg must be a type",
+        ):
+            Concatenate[(str,), P]
```

File: src/test_typing_extensions.py [hunk 1]
```diff
@@ -1720,12 +1720,15 @@ class C(Generic[T]): pass
...
-                         (Concatenate[int, P], int))
         self.assertEqual(get_args(Required[int]), (int,))
         self.assertEqual(get_args(NotRequired[int]), (int,))
         self.assertEqual(get_args(Unpack[Ts]), (Ts,))
         self.assertEqual(get_args(Unpack), ())
+        self.assertEqual(get_args(Callable[Concatenate[int, P], int]),
+                         (Concatenate[int, P], int))
+        with self.subTest("Concatenate[int, ...]"):
+            self.assertEqual(get_args(Callable[Concatenate[int, ...], int]),
+                        (Concatenate[int, ...], int))
```
Comment (JelleZijlstra) [inline src/test_typing_extensions.py:1731]:
```suggestion
        self.assertEqual(get_args(Callable[Concatenate[int, ...], int]),
                        (Concatenate[int, ...], int))
```
No need for the subtest

File: src/test_typing_extensions.py [hunk 2]
```diff
@@ -5371,13 +5383,27 @@ def test_valid_uses(self):
         self.assertEqual(C1.__origin__, C2.__origin__)
         self.assertNotEqual(C1, C2)
 
-        # Test collections.abc.Callable too.
-        if sys.version_info[:2] >= (3, 9):
-            C3 = collections.abc.Callable[Concatenate[int, P], int]
-            C4 = collections.abc.Callable[Concatenate[int, T, P], T]
+        with self.subTest("typing.Callable with Ellipsis"):
```
Comment (JelleZijlstra) [inline src/test_typing_extensions.py:5386]: Perhaps better to make a new test method.

File: src/test_typing_extensions.py [hunk 3]
```diff
@@ -5390,25 +5416,50 @@ def test_invalid_uses(self):
...
+    def test_invalid_use(self):
+        # Assure that `_type_check` is called.
+        P = ParamSpec('P')
+        with self.assertRaisesRegex(
+            TypeError,
+            "each arg must be a type",
+        ):
+            Concatenate[(str,), P]
+
+    @skipUnless(TYPING_3_11_0 or (3, 10, 0) <= sys.version_info < (3, 10, 2),
```
Comment (JelleZijlstra) [inline src/test_typing_extensions.py:5438]: In this case I think I'd want to block it on all 3.10 versions. It's a bad experience if something works on early 3.10 and breaks if people upgrade to a later bugfix release. Better to allow it only on 3.11+.

Comment (JelleZijlstra) [reply src/test_typing_extensions.py:5438]: Also, the version check seems wrong here, the change was made in 3.10.3 not 3.10.2.

File: src/typing_extensions.py [hunk 4]
```diff
@@ -1795,28 +1795,56 @@ def __parameters__(self):
             return tuple(
                 tp for tp in self.__args__ if isinstance(tp, (typing.TypeVar, ParamSpec))
             )
+# 3.10+
+else:
+    _ConcatenateGenericAlias = typing._ConcatenateGenericAlias
 
+# 3.8-3.9.2
+class _EllipsisDummy: ...
```
Comment (JelleZijlstra) [inline src/typing_extensions.py:1803]: If you make this inherit from ParamSpec, can it fix 3.10.3+?

Comment (Daraan) [reply src/typing_extensions.py:1803]: I might be missing something here. To what case are you referring here?
This backport of `Concatenate[...]` also works for all 3.10, as the critical part is in `_concatenate_getitem`

Is it possibly related to the following test I jointly introduced also in #479?

```python
def test_alias_subscription_with_ellipsis(self):
    P = ParamSpec('P')
    X = Callable[Concatenate[int, P], Any]
    C1 = X[...]  # <-- Needs PR #479 for 3.10.3+
    self.assertEqual(get_args(C1), (Concatenate[int, ...], Any))  # <-- needs this PR
```

---

That aside; It is possible to do `_EllipsisDummy = ParamSpec("_EllipsisDummy")` if needed as subclassing `ParamSpec` is (currently) not possible. I thought using a new class might have less side-effects than using a ParamSpec instance.

Comment (JelleZijlstra) [reply src/typing_extensions.py:1803]: I'm referring to the comment right above here, about the test that works only in 3.10.0-3.10.2 but not 3.10.3+. Making `_EllipsisDummy` an instance of ParamSpec might fix that test for later versions of 3.10.

File: src/test_typing_extensions.py (hunk 2 above)
Comment (JelleZijlstra) [reply src/test_typing_extensions.py:5386]: Sure!

File: src/typing_extensions.py (hunk 4 above)
Comment (Daraan) [reply src/typing_extensions.py:1803]: I am sorry for the confusion. I do think we talk about the same location :), if not correct me please. I made the necessary edits above.

#479 allows to backport that test to all versions of 3.10, however for #479 the test `self.assertEqual(get_args(C1), (Concatenate[int, ...], Any))` would need this PR to work without an error on 3.10 as well.  
Both PRs are needed make this test work for the 3.10 version range.

Comment (JelleZijlstra) [reply src/typing_extensions.py:1803]: I see, thanks. #479 is merged now, so would you mind updating this PR? There's a merge conflict right now.

Comment (Daraan) [reply src/typing_extensions.py:1822]: Sure. Unified the code and the test we talked about, aside from that added one more test for invalid usage.

**Review by JelleZijlstra — APPROVED** [review]

**Pull request merged**